### PR TITLE
Add a parser for ODL feeds serialized using OPDS 2.x

### PR DESCRIPTION
**Pull request by jonathangreen**
This PR is moved from here: 
https://github.com/lyrasis/simplye-webpub-manifest-parser/pull/2

This PR adds enhances functionality of `webpub-manifest-parser`:
1. It adds a parser for ODL feeds serialized using OPDS 2.x
2. It adds error recovery logic: instead of raising an exception for the first error found, the parsers tries to record all of them in `ManifestParserResult.errors`.
3. It adds the following integration tests:
- OPDS 2.x: 
  - Feedbooks, Fulcrum, 
  - ProQuest (skipped by default - requires a running SOCKS proxy since the ProQuest feed is accessible only by whitelisted IP addresses).
- ODL 2.x: 
  - DPLA Exchange (skipped by default - requires credentials to pull the feed).

**NOTE: This new version of `webpub-manifest-parser` is not backward compatible with [NYPL's version](https://github.com/NYPL-Simplified/python-webpub-manifest-parser) and it must NOT be uploaded to [NYPL's PyPI repo](https://pypi.org/project/webpub-manifest-parser/) because it will break NYPL's CM.**

### CODE REVIEW

**Comment by jonathangreen** [comment]
@vbessonov mind merging main into this so we get tests running on it?

**Comment by vbessonov** [comment]
@jonathangreen, it seems like GitHub has time restrictions for running tests: [the OPDS 2.x integration tests](https://github.com/ThePalaceProject/webpub-manifest-parser/blob/feature/odl2/tests/webpub_manifest_parser/opds2/test_production_feeds.py#L11) time out. 
Do you think it's better to skip them? They're quite useful but they access production Feedbooks and Fulcrum feeds.

**Comment by jonathangreen** [comment]
@vbessonov how long do the integration tests take to run? It does seem useful to have them. Its possible we could run them as a separate job from the unit tests. Are you running the tests in your fork of the repo, is that where you see the job timing out?

**Comment by vbessonov** [comment]
For example, [this run](https://github.com/ThePalaceProject/webpub-manifest-parser/pull/2/checks?check_run_id=2601417379) failed with the error `The operation was canceled` after 5 seconds.

**Comment by jonathangreen** [comment]
I don't think the tests are timing out if they only take 20s. When one test fails it cancels others instead of keeping them running. I'm wondering if that is what happened. The last test run it looks like most of the tests passed. Except for python 3.7. Rerunning to see if they will pass. 

The output of the tests seems really verbose. Trying to load the logs in firefox is crashing my browser. I wonder if we should be running the tests without such verbose logging.

**Comment by jonathangreen** [comment]
Downloaded the log files from github. It looks like each run is producing 43MB of logs 😳. 

Did manage to find this in there though:
```
2021-05-17T14:57:32.3511902Z ERROR: test (tests.webpub_manifest_parser.odl.test_parser.ODLParserTest)
2021-05-17T14:57:32.3513488Z ----------------------------------------------------------------------
2021-05-17T14:57:32.3514143Z Traceback (most recent call last):
2021-05-17T14:57:32.3516648Z   File "/home/runner/work/webpub-manifest-parser/webpub-manifest-parser/tests/webpub_manifest_parser/odl/test_parser.py", line 22, in test
2021-05-17T14:57:32.3517994Z     result = parser.parse_file(input_file_path)
2021-05-17T14:57:32.3519600Z   File "/home/runner/work/webpub-manifest-parser/webpub-manifest-parser/src/webpub_manifest_parser/core/__init__.py", line 87, in parse_file
2021-05-17T14:57:32.3520750Z     return self._parse(manifest_json)
2021-05-17T14:57:32.3522198Z   File "/home/runner/work/webpub-manifest-parser/webpub-manifest-parser/src/webpub_manifest_parser/core/__init__.py", line 62, in _parse
2021-05-17T14:57:32.3523449Z     manifest = self._syntax_analyzer.analyze(manifest_json)
2021-05-17T14:57:32.3525011Z   File "/home/runner/work/webpub-manifest-parser/webpub-manifest-parser/src/webpub_manifest_parser/core/syntax.py", line 335, in analyze
2021-05-17T14:57:32.3526281Z     manifest = self._parse_object(manifest_json, manifest.__class__)
2021-05-17T14:57:32.3527882Z   File "/home/runner/work/webpub-manifest-parser/webpub-manifest-parser/src/webpub_manifest_parser/core/syntax.py", line 315, in _parse_object
2021-05-17T14:57:32.3529105Z     self._set_non_scalar_value(json_content, ast_object)
2021-05-17T14:57:32.3530698Z   File "/home/runner/work/webpub-manifest-parser/webpub-manifest-parser/src/webpub_manifest_parser/core/syntax.py", line 246, in _set_non_scalar_value
2021-05-17T14:57:32.3531899Z     property_value, object_property
2021-05-17T14:57:32.3533421Z   File "/home/runner/work/webpub-manifest-parser/webpub-manifest-parser/src/webpub_manifest_parser/core/syntax.py", line 134, in _parse_nested_object
2021-05-17T14:57:32.3534750Z     processed_item = self._parse_object(item, type_parser.type)
2021-05-17T14:57:32.3536357Z   File "/home/runner/work/webpub-manifest-parser/webpub-manifest-parser/src/webpub_manifest_parser/core/syntax.py", line 315, in _parse_object
2021-05-17T14:57:32.3537565Z     self._set_non_scalar_value(json_content, ast_object)
2021-05-17T14:57:32.3539133Z   File "/home/runner/work/webpub-manifest-parser/webpub-manifest-parser/src/webpub_manifest_parser/core/syntax.py", line 250, in _set_non_scalar_value
2021-05-17T14:57:32.3540475Z     ast_object, object_property_name, object_property, property_value
2021-05-17T14:57:32.3542145Z   File "/home/runner/work/webpub-manifest-parser/webpub-manifest-parser/src/webpub_manifest_parser/core/syntax.py", line 284, in _set_property_value
2021-05-17T14:57:32.3543711Z     property_value = object_property.parser.parse(property_value)
2021-05-17T14:57:32.3545378Z   File "/home/runner/work/webpub-manifest-parser/webpub-manifest-parser/src/webpub_manifest_parser/core/parsers.py", line 586, in parse
2021-05-17T14:57:32.3546539Z     if self._unique_items and item in seen:
2021-05-17T14:57:32.3547371Z TypeError: unhashable type: 'ODLLicense'
```

Looks like maybe this is still failing for Python 3 due to the hashing changes.

**Comment by vbessonov** [comment]
Thank you, I also managed to figure it out by myself. I'll make new AST types hashable and also increase the logging level to `WARNING` to eliminate such verbose output.

**Review by tdilauro — COMMENTED** [review]
@vbessonov This looks great. In addition to the new functionality, I appreciate the attention you've given to naming.

I have a couple of comments:
- There are a bunch of error messages that are duplicated throughout the code. There are a couple examples in my inline comments, but there are others. I don't think this should prevent merging this PR, but it would be nice to address this in a future PR to make the code more maintainable.
- I think this is a very useful package and could see it being used in multiple projects. For that reason, I'm wondering if we should find another home entirely for the integration tests. @jonathangreen I'd appreciate your thoughts here, as well, especially since you've weighed in on the utility of those tests.

File: tests/webpub_manifest_parser/opds2/test_production_feeds.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                "Feedbooks",
+                "https://catalog.feedbooks.com/catalog/index.json",
+            ),
+            (
+                "fulcrum",
+                "https://www.fulcrum.org/api/opds/umpebc_oa",
+            ),
+        ]
+    )
+    def test_production_feed(self, _, url, encoding="utf-8"):
```
[the remark this answers is not in the document]
Comment (tdilauro) [reply tests/webpub_manifest_parser/opds2/test_production_feeds.py:24]: The PR description mentions:
> OPDS 2.x: Feedbooks, Fulcrum, ProQuest (skipped by default - requires a running SOCKS proxy since the ProQuest feed is accessible only by whitelisted IP addresses).

The wording is ambiguous there. Feedbooks and Fulcrum are not skipped by default, only ProQuest, so it might be helpful to reword this to make it more clear.

File: tests/webpub_manifest_parser/odl/test_production_feeds.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+            for index, error in enumerate(feed_parsing_result.errors):
+                print("{0}. {1}".format(index + 1, error.error_message))
+
+            print("")
+
+    @parameterized.expand(
+        [
+            (
+                "New Market OPDS 2.0 + ODL feed",
```
[the remark this answers is not in the document]
Comment (vbessonov) [reply tests/webpub_manifest_parser/odl/test_production_feeds.py]: I updated the test case's name.

File: tests/webpub_manifest_parser/opds2/test_production_feeds.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vbessonov) [reply tests/webpub_manifest_parser/opds2/test_production_feeds.py:24]: I updated the description.

File: tests/webpub_manifest_parser/rwpm/test_syntax.py [hunk 3]
```diff
@@ -14,14 +15,15 @@
     PresentationMetadata,
     Subject,
 )
-from webpub_manifest_parser.core.parsers import DocumentParser, ValueParsingError
+from webpub_manifest_parser.core.parsers import ValueParserError
```
[the remark this answers is not in the document]
Comment (vbessonov) [reply tests/webpub_manifest_parser/rwpm/test_syntax.py]: You're right, it's not used in this module, I removed it.

File: src/webpub_manifest_parser/core/ast.py [hunk 4]
```diff
@@ -857,11 +1038,13 @@ def __init__(self, key, required, role, collection_class=CompactCollection):
         """
         if not isinstance(role, CollectionRole):
             raise ValueError(
-                "Role argument must be an instance of CollectionRole class"
+                "Argument 'role' must be an instance of {0}".format(CollectionRole)
             )
```
[the remark this answers is not in the document]
Comment (vbessonov) [reply src/webpub_manifest_parser/core/ast.py:1042]: I'm just afraid that moving the validation logic to separate methods will make the code harder to understand.
We would need to use [decorators](https://hackaday.com/2018/08/31/an-introduction-to-decorators-in-python/) instead.

File: tests/webpub_manifest_parser/opds2/test_production_feeds.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vbessonov) [reply tests/webpub_manifest_parser/opds2/test_production_feeds.py:24]: Regarding the skipping the test - this is what I'm not sure about. I would like to have these tests running all the time since they allow to understand if the parser is able to parse correct feeds. But at the same time those tests can give us false negatives: they can fail because of errors in the feeds or network issues.

[the remark this answers is not in the document]
Comment (vbessonov) [reply tests/webpub_manifest_parser/opds2/test_production_feeds.py:25]: I skipped them.

File: src/webpub_manifest_parser/core/ast.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tdilauro) [reply src/webpub_manifest_parser/core/ast.py:1042]: A good first step for a future PR would be to consolidate the error text, but it doesn't need to be in this PR.

File: tests/webpub_manifest_parser/opds2/test_production_feeds.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tdilauro) [reply tests/webpub_manifest_parser/opds2/test_production_feeds.py:25]: > Regarding the skipping the test - this is what I'm not sure about. I would like to have these tests running all the time since they allow to understand if the parser is able to parse correct feeds. But at the same time those tests can give us false negatives: they can fail because of errors in the feeds or network issues.

Exactly. I would say that those feeds represent "real" feeds, but not necessarily "correct" feeds. Additionally, what constitutes a "correct" feed may change over time as the spec changes.

My personal opinion is that these are separate problems:
- We can validate correct parsing of "correct" feeds with unit tests and a suite of correct feeds.
- We can assure that we are able to process the applicable "real" feeds with integration tests.

For us, applicable feeds would be the feeds that we use, but another user of this package might be using different feeds. As you've noted, another issue for the integration tests is the need for credentials. That might argue for situating these tests in an environment where we have these credentials.

[the remark this answers is not in the document]
Comment (jonathangreen) [reply tests/webpub_manifest_parser/opds2/test_production_feeds.py:25]: I tend to agree @tdilauro. I like the idea of storing a "correct" version of these feeds as a test fixture and running a test against it. 

Then as future work we could add a separate integration test action that runs integration tests against interesting feeds, so we can see if something changes in the integration that causes these tests to start failing. This could be either done here in github or in our hosting environment.

[the remark this answers is not in the document]
Comment (vbessonov) [reply tests/webpub_manifest_parser/opds2/test_production_feeds.py:25]: @jonathangreen, @tdilauro, is that right that I can just grab a piece of Feedbooks and Fulcrum feeds and save those snippets in this repo? Won't it be a violation of any copyrights?

[the remark this answers is not in the document]
Comment (tdilauro) [reply tests/webpub_manifest_parser/opds2/test_production_feeds.py:25]: I suspect that the feeds themselves are not copyrightable; though, that might not be true for the parts of them that involve some intellectual (non-factual) contribution (e.g., `.publications[].metadata.description`).

For what I called "correct" feeds above, I think there are a couple of ways forward:
- I don't think we need to use the exact content of specific feeds to create fixtures. We could, for example, change links.
- We could also seek permission to use them for this purpose.
- We could create our own or acquire sample feeds.

Dealing with validation of feeds in which we have specific production interest could be handled separately in integration tests. The integration would need to publish the feeds, so this would not be an issue.

[the remark this answers is not in the document]
Comment (vbessonov) [reply tests/webpub_manifest_parser/opds2/test_production_feeds.py:25]: There is already a [snippet](https://github.com/ThePalaceProject/webpub-manifest-parser/blob/feature/odl2/tests/files/odl/feed.json) in the codebase. I took it from the ODL spec and it's quite illustrative.

[the remark this answers is not in the document]
Comment (tdilauro) [reply tests/webpub_manifest_parser/opds2/test_production_feeds.py:25]: Yeah, that seems fine. We could update that or add more later, if needed.

[the remark this answers is not in the document]
Comment (vbessonov) [reply tests/webpub_manifest_parser/opds2/test_production_feeds.py:25]: @jonathangreen, @tdilauro can we merge this PR?

**Review by tdilauro — APPROVED** [review]
Yes. This looks good to me.

I don't see any problem with leaving the skipped integration tests in, at least for now.

File: tests/webpub_manifest_parser/opds2/test_production_feeds.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonathangreen) [reply tests/webpub_manifest_parser/opds2/test_production_feeds.py:25]: Looks good to me as well.
